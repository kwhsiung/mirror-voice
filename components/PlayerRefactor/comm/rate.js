export const ratesAvailable = [2.0, 1.5, 1.0, 0.75, 0.5]
