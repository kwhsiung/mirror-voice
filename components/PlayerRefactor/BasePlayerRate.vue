<template>
  <div
    v-click-outside="hidePicker"
    class="rate"
  >
    <button
      class="button-toggler"
      @click="togglePicker"
      v-text="playbackRateCurrentIndicator"
    />
    <ol
      v-show="showPicker"
      class="picker rate__picker"
    >
      <li
        v-for="rate in ratesAvailable"
        :key="rate"
        :class="[
          'picker__list-item',
          { 'active': rate === playbackRateCurrent }
        ]"
        @click="pickRate(rate)"
        v-text="formatRateDisplay(rate)"
      />
    </ol>
  </div>
</template>

<script>
import _ from 'lodash'
import { ratesAvailable } from './comm/rate'

export default {
  props: {
    playbackRateCurrent: {
      type: Number,
      default: 1.0,
      validator(value) {
        return ratesAvailable.includes(value)
      }
    }
  },
  data() {
    return {
      ratesAvailable,
      showPicker: false
    }
  },
  computed: {
    playbackRateCurrentIndicator() {
      const playbackRateFloat = this.formatRateDisplay(this.playbackRateCurrent)
      return `X ${playbackRateFloat}`
    }
  },
  methods: {
    togglePicker() {
      this.showPicker = !this.showPicker
    },
    hidePicker() {
      this.showPicker = false
    },
    formatRateDisplay(rate) {
      const rateString = rate.toString()
      return _.padEnd(rateString, 3, '.0')
    },
    pickRate(rate) {
      this.$emit('update:playbackRateCurrent', rate)
    }
  }
}
</script>

<style lang="stylus" scoped>
.rate
  height 100%
  display flex
  justify-content center
  align-items center
  position relative
  &__picker
    position absolute
    bottom calc(100% + 3px)
    width 52px
    background-color #313131
    border-radius 2px

.button-toggler
  min-width 57px
  font-size 18px
  font-weight 600
  color #e5e5e5
  line-height 1
  cursor pointer
  user-select none
  background-color transparent
  border none
  padding 0
  outline none

.picker
  list-style none
  margin 0
  padding 20px 13px
  &__list-item
    color white
    line-height 1.29
    cursor pointer
    &.active
      color #d84939
    & + &
      margin 17px 0 0 0

@media (max-width 768px)
  .button-toggler
    min-width 50px
    font-size 15px
    text-align left
</style>
