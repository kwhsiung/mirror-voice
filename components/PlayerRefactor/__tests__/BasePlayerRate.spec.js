import { shallowMount } from '@vue/test-utils'
import BasePlayerRate from '../BasePlayerRate.vue'
import { ratesAvailable } from '../comm/rate'

describe('BasePlayerRate', () => {
  test('render "X 0.75" text in toggler while playbackRateCurrent props is 0.75', () => {
    const playbackRateCurrent = 0.75
    const wrapper = shallowMount(BasePlayerRate, {
      propsData: {
        playbackRateCurrent
      }
    })
    const button = wrapper.find('.button-toggler')

    expect(button.text()).toBe(`X ${playbackRateCurrent.toString()}`)
  })

  test('has "active" class on list item of current playbackrate', () => {
    const playbackRateCurrentIndex = 1
    const playbackRateCurrent = ratesAvailable[playbackRateCurrentIndex]
    const wrapper = shallowMount(BasePlayerRate, {
      propsData: {
        playbackRateCurrent
      }
    })

    const listItems = wrapper.findAll('li')
    const listItemAcitve = listItems.at(playbackRateCurrentIndex)

    expect(listItemAcitve.classes()).toContain('active')

    const listItemsNotActive = listItems.filter(
      w => !w.classes().includes('active')
    )

    expect(listItemsNotActive.length).toBe(ratesAvailable.length - 1)
  })

  test('emit "update:playbackRateCurrent" with different rate value while rate list item clicked', () => {
    const playbackRateChangedIndex = 1
    const playbackRateChanged = ratesAvailable[playbackRateChangedIndex]
    const wrapper = shallowMount(BasePlayerRate)

    const listItem = wrapper.findAll('li').at(playbackRateChangedIndex)
    listItem.trigger('click')

    expect(wrapper.emitted('update:playbackRateCurrent')[0]).toEqual([
      playbackRateChanged
    ])
  })
})

describe('snapshot tests', () => {
  test('renders component correctly', () => {
    const wrapper = shallowMount(BasePlayerRate)

    expect(wrapper.element).toMatchSnapshot()
  })
})
