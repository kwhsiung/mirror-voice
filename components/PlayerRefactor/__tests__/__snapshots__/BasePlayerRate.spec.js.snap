// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests renders component correctly 1`] = `
<div
  class="rate"
>
  <button
    class="button-toggler"
  >
    X 1.0
  </button>
   
  <ol
    class="picker rate__picker"
    style="display: none;"
  >
    <li
      class="picker__list-item"
    >
      2.0
    </li>
    <li
      class="picker__list-item"
    >
      1.5
    </li>
    <li
      class="picker__list-item active"
    >
      1.0
    </li>
    <li
      class="picker__list-item"
    >
      0.75
    </li>
    <li
      class="picker__list-item"
    >
      0.5
    </li>
  </ol>
</div>
`;
